Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewDelegator.cs

```diff
@@ -10,15 +11,62 @@ public class ItemsViewDelegator<TItemsView, TViewController> : UICollectionViewD
 		where TItemsView : ItemsView
 		where TViewController : ItemsViewController<TItemsView>
 	{
+
+		IDictionary<Type, CGSize> _sizeChache = new Dictionary<Type,CGSize>();
 		public ItemsViewLayout ItemsViewLayout { get; }
 		public TViewController ViewController { get; }
+		public IItemsViewSource ItemsSource { get; }
 
 		protected float PreviousHorizontalOffset, PreviousVerticalOffset;
 
-		public ItemsViewDelegator(ItemsViewLayout itemsViewLayout, TViewController itemsViewController)
+		public ItemsViewDelegator(ItemsViewLayout itemsViewLayout, TViewController itemsViewController, IItemsViewSource itemsViewSource)
 		{
 			ItemsViewLayout = itemsViewLayout;
 			ViewController = itemsViewController;
+			ItemsSource = itemsViewSource;
+		}
+
+
+		/// <summary>
+		/// Concretions have to override this method as this only sizes the items in a vertical and horizontal collectionview
+		/// Per default this method returns the Estimated size when its not overriden.
+		/// This method is called before the rendering process and sizes the cell correctly before it is displayed in the collectionview
+		/// Calling the base implementation of this method will throw an exception when overriding the method
+		/// </summary>
+		/// <returns></returns>
+		public override CGSize GetSizeForItem(UICollectionView collectionView, UICollectionViewLayout layout, NSIndexPath indexPath)
+		{
+			if (ItemsViewLayout.ItemSizingStrategy == ItemSizingStrategy.MeasureFirstItem)
+				return ItemsViewLayout.EstimatedItemSize;
+
+			var vm = ItemsSource[indexPath];
+
+			// try to get a cached size
+			if (_sizeChache.ContainsKey(vm.GetType()))
+				return _sizeChache[vm.GetType()];
```

Comment:
This can still be improved by using a `IDictionary<IndexPath, CGSize>` as cache.

And whenever the item changes in size we would have to update the value at the given index path.

however, there is still a catch: the method which check for a layout invalidation is not located in the delegate.
it is located in either the `ItemsViewController` or in the `ItemsViewLayout` (i cant rememebr right now and im rly in hurry. will maybe update this later or work on it)

This means that the _cache has to be accessible in that class as well.
But then this will be a perfectly working solution.